[JENA-1911]  vue

kinow opened the pull request:
:clapper: take 2 !

@afs finding myself now in a short sabbatical with extra spare time, how about we resuscitate the idea to revamp the UI #781? Doesn't need to be Vue, we can take another direction and do it incrementally, or use React, etc. WDYT?

Bruno

afs commented:
Hi @kinow

This is what I tried:

Pulled the PR and did "mvn install" . Unpacked the `apache-jena-fuseki` zip and run `java -jar fuseki-server.jar`.

I get new appearance but the UI does not seem to be able to talk to the server. The server status is always red, and creating a dataset fails.

There doesn't seem to be a "run" directory. What have I missed?

kinow commented:
Hi @afs 

That's good progress I guess. Last I worked on the branch, I wasn't sure how to package the new UI with the old Servlets project. Looks like the build is at least working partially. I also got the new UI, but it's not pointing to the right backend URL. I'm taking a look and trying to get it to work building with `mvn install` and using the ZIP as you did :+1:

kinow commented:
Hmm, running `FusekiCmd` in Eclipse it works fine.

![image](https://user-images.githubusercontent.com/304786/143190685-e8f79acf-3b48-497e-b442-2bbf8b665316.png)

![image](https://user-images.githubusercontent.com/304786/143190710-6eac5c1c-3538-4617-a5ad-290aed6a8a8e.png)

So it must be something with how I tried to glue the Vue UI in the Fuseki war/jar.

kinow commented on jena-fuseki2/jena-fuseki-webapp/src/main/java/org/apache/jena/fuseki/cmd/JettyFusekiWebapp.java:
```diff
@@ -77,7 +77,7 @@
     // Standalone jar
     public static final String resourceBase1   = "webapp";
```
Right, I think the ZIP file will call the standalone jar (when used with `java -jar fuseki-server.jar`). So this `webapp` will be used.

When using in Eclipse I had to modify the line below to load the new UI.

kinow commented:
Found it @afs. I had a `webapp` folder in the ZIP file (which is searched/used first by the command if it contains a `web.xml`).

However, the webapp folder had the contents of the new UI only. `jena-fuseki-webapp` has `jena-fuseki-ui` as dependency, and has a `web.xml` plus copies the `jena-fuseki-ui` new files as well.

So the solution is the last commit, where I copied the webapp generated by `jena-fuseki-webapp` into the assembly output. It should work now @afs :crossed_fingers: If that works, you can try resizing the browser window, or using a mobile. If I recall correctly, I did a few adjustments in the CSS/HTML to support responsive UI in most screens.

-Bruno

kinow commented on jena-fuseki2/apache-jena-fuseki/assembly-dist.xml:
```diff
@@ -103,7 +103,7 @@
 
     <fileSet>
-      <directory>../jena-fuseki-webapp/src/main/webapp</directory>
+      <directory>../jena-fuseki-webapp/target/webapp/</directory>
```
The `target` contains the `jena-fuseki-webapp` files, plus the dist files of `jena-fuseki-ui` (see `webapp`'s `pom.xml`).

rvesse commented on jena-fuseki2/jena-fuseki-ui/pom.xml:
```diff
@@ -0,0 +1,98 @@
+                        </goals>
+                        <configuration>
+                        	<nodeVersion>${node.version}</nodeVersion>
+                            <yarnVersion>${yarn.version}</yarnVersion>
```
Mismatched indentation here

kinow replied:
Fixed, thanks @rvesse !

rvesse commented on jena-fuseki2/jena-fuseki-war/pom.xml:
```diff
@@ -33,12 +33,20 @@
 
   <dependencies>
 
+	<!-- backend servlets, filters (CORS, Shiro, etc.) -->
```
Tabs vs Spaces?

kinow replied:
Fixed, thanks @rvesse !

rvesse commented on jena-fuseki2/jena-fuseki-webapp/pom.xml:
```diff
@@ -72,6 +72,13 @@
     </dependency>
 
+	<!-- frontend assets (HTML, JS, CSS, images, etc.) -->
+	<dependency>
```
Mismatched indentation

kinow replied:
Fixed, thanks @rvesse !

rvesse commented on jena-fuseki2/jena-fuseki-war/pom.xml:
```diff
@@ -62,14 +70,19 @@
         </executions>
         <configuration>
-          <warSourceDirectory>../jena-fuseki-webapp/src/main/webapp</warSourceDirectory>
+          <warSourceDirectory>../jena-fuseki-ui/target/dist/</warSourceDirectory>
```
Shouldn't this be `jena-fuseki-webapp/target/webapp/`

If've been doing some questionable things with Maven and WARs in $dayjob lately and even I'm having trouble following this

kinow replied:
@rvesse 

>Shouldn't this be jena-fuseki-webapp/target/webapp/

Ah! You are correct. Yesterday I jotted down a quick diagram of the new build hierarchy, and the `webapp` module now includes the contents of the `ui` module plus some other files.

![20211125_112226](https://user-images.githubusercontent.com/304786/143320379-59f9c409-2c92-4cc4-90ed-2c521865f0eb.jpg)

>If've been doing some questionable things with Maven and WARs in $dayjob lately and even I'm having trouble following this

:rofl:

kinow commented:
Fixed code based on @rvesse 's feedback (thanks) and tested the zip again after a `mvn clean install -Pdev`. Tests passing, UI working OK, and successfully created a dataset.

afs commented:
Looks great! I got it work with Fuseki Main (via https://github.com/afs/fuseki-mod-ui) serving the files from the classpath by copying to src/main/resources/vue and serving that from the classpath with Jetty static assets servlet.

1. What source is copied from other open source projects that will end up? (For the LICENSE / NOTICE files)
1. As it depends on no Jena artifacts, can we move the build to just after `jena-fuseki-access`? This makes it "one of the modules" and can then go into Fuseki Mian as easily as Fuseki webapp.
1. How easy is to to provide a version with no "manage" version?
1. Formatting of results - can have have "" on strings and _: for bnode labels (I realise this is about yasgui/yasr)
1. pom.xml - 2 spaces indent!

About 3: if the files can be served from the classpath, then Fuseki Main can have query/data load without a fixed directory structure on the server side because all the configuration is from the command line (or an assembler file on the command line).
The on-disk state is one of the complicated bits of the original webapp.

namedgraph commented:
Hi,

So this is not a Java webapp (WAR packaging)? Because in a WAR static files
(.js, .css etc.) should go into ‘webapp’, not ‘resources’.

On Sat, 27 Nov 2021 at 19.24, Andy Seaborne ***@***.***>
wrote:

> Looks great! I got it work with Fuseki Main (via
> https://github.com/afs/fuseki-mod-ui) serving the files from the
> classpath by copying to src/main/resources/vue and serving that from the
> classpath with Jetty static assets servlet.
>
>    1. What source is copied from other open source projects that will end
>    up? (For the LICENSE / NOTICE files)
>    2. As it depends on no Jena artifacts, can we move the build to just
>    after jena-fuseki-access? This makes it "one of the modules" and can
>    then go into Fuseki Mian as easily as Fuseki webapp.
>    3. How easy is to to provide a version with no "manage" version?
>    4. Formatting of results - can have have "" on strings and _: for
>    bnode labels (I realise this is about yasgui/yasr)
>    5. pom.xml - 2 spaces indent!
>
> About 3: if the files can be served from the classpath, then Fuseki Main
> can have query/data load without a fixed directory structure on the server
> side because all the configuration is from the command line (or an
> assembler file on the command line).
> The on-disk state is one of the complicated bits of the original webapp.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/1115#issuecomment-980780893>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGPM5RWEYO3A55NAJBRMBDUOEO6HANCNFSM5IQPNC4A>
> .
>

afs commented:
@namedgraph  -- I was trying the UI with Fuseki Main, not Fusekji Webapp.

See the build in the PR for `jena-fuseki-webapp` for locations in that packaging.

kinow commented:
>Looks great! I got it work with Fuseki Main (via https://github.com/afs/fuseki-mod-ui) serving the files from the classpath by copying to src/main/resources/vue and serving that from the classpath with Jetty static assets servlet.

Great! :tada: 

>What source is copied from other open source projects that will end up? (For the LICENSE / NOTICE files)

Hmmmm, that is a very good question. Here's the list of dependencies:

<details>
<summary>`yarn list --prod`</summary>

```bash
yarn list v1.22.5
├─ @fortawesome/fontawesome-common-types@0.2.35
├─ @fortawesome/fontawesome-svg-core@1.2.35
│  └─ @fortawesome/fontawesome-common-types@^0.2.35
├─ @fortawesome/free-solid-svg-icons@5.15.3
│  └─ @fortawesome/fontawesome-common-types@^0.2.35
├─ @fortawesome/vue-fontawesome@2.0.2
├─ @nuxt/opencollective@0.3.2
│  ├─ chalk@^4.1.0
│  ├─ consola@^2.15.0
│  └─ node-fetch@^2.6.1
├─ @triply/yasgui-utils@4.2.4
│  ├─ @types/node@14.0.23
│  └─ store@^2.0.12
├─ @triply/yasqe@4.2.7
│  ├─ @triply/yasgui-utils@^4.2.4
│  ├─ @types/lodash-es@^4.17.3
│  ├─ codemirror@^5.51.0
│  ├─ lodash-es@^4.17.15
│  ├─ query-string@^6.10.1
│  └─ superagent@5.3.1
├─ @triply/yasr@4.2.8
│  ├─ @fortawesome/fontawesome-common-types@0.2.30
│  ├─ @fortawesome/free-solid-svg-icons@^5.14.0
│  ├─ @fortawesome/free-solid-svg-icons@5.14.0
│  │  └─ @fortawesome/fontawesome-common-types@^0.2.30
│  ├─ @triply/yasgui-utils@^4.2.4
│  ├─ @triply/yasqe@^4.2.7
│  ├─ @types/lodash-es@^4.17.3
│  ├─ codemirror@^5.51.0
│  ├─ colors@^1.4.0
│  ├─ column-resizer@^1.3.4
│  ├─ datatables.net-dt@^1.10.20
│  ├─ datatables.net@^1.10.20
│  ├─ dompurify@^2.0.7
│  ├─ jquery@^3.5.0
│  ├─ json2csv@^5.0.1
│  ├─ lodash-es@^4.17.15
│  └─ n3@^1.3.5
├─ @types/color-name@1.1.1
├─ @types/lodash-es@4.17.3
│  └─ @types/lodash@*
├─ @types/lodash@4.14.159
├─ @types/node@14.0.23
├─ ansi-styles@4.2.1
│  ├─ @types/color-name@^1.1.1
│  └─ color-convert@^2.0.1
├─ asynckit@0.4.0
├─ axios@0.21.1
│  └─ follow-redirects@^1.10.0
├─ bootstrap-vue@2.21.2
│  ├─ @nuxt/opencollective@^0.3.2
│  ├─ bootstrap@>=4.5.3 <5.0.0
│  ├─ popper.js@^1.16.1
│  ├─ portal-vue@^2.1.7
│  └─ vue-functional-data-merge@^3.1.0
├─ bootstrap@4.6.0
├─ chalk@4.1.0
│  ├─ ansi-styles@^4.1.0
│  └─ supports-color@^7.1.0
├─ codemirror@5.56.0
├─ color-convert@2.0.1
│  └─ color-name@~1.1.4
├─ color-name@1.1.4
├─ colors@1.4.0
├─ column-resizer@1.3.5
│  └─ string-hash@~1.1.3
├─ combined-stream@1.0.8
│  └─ delayed-stream@~1.0.0
├─ commander@5.1.0
├─ component-emitter@1.3.0
├─ consola@2.15.3
├─ cookiejar@2.1.2
├─ core-js@3.9.1
├─ datatables.net-dt@1.10.21
│  ├─ datatables.net@1.10.21
│  └─ jquery@>=1.7
├─ datatables.net@1.10.21
│  └─ jquery@>=1.7
├─ debug@4.1.1
│  └─ ms@^2.1.1
├─ decode-uri-component@0.2.0
├─ delayed-stream@1.0.0
├─ dompurify@2.0.12
├─ fast-safe-stringify@2.0.7
├─ follow-redirects@1.13.0
├─ form-data@3.0.0
│  ├─ asynckit@^0.4.0
│  ├─ combined-stream@^1.0.8
│  └─ mime-types@^2.1.12
├─ formidable@1.2.2
├─ has-flag@4.0.0
├─ inherits@2.0.4
├─ jquery@3.5.1
├─ json2csv@5.0.1
│  ├─ commander@^5.0.0
│  ├─ jsonparse@^1.3.1
│  └─ lodash.get@^4.4.2
├─ jsonparse@1.3.1
├─ lodash-es@4.17.15
├─ lodash.get@4.4.2
├─ methods@1.1.2
├─ mime-db@1.44.0
├─ mime-types@2.1.27
│  └─ mime-db@1.44.0
├─ mime@2.4.6
├─ ms@2.1.2
├─ n3@1.6.2
│  ├─ queue-microtask@^1.1.2
│  └─ readable-stream@^3.6.0
├─ node-fetch@2.6.1
├─ popper.js@1.16.1
├─ portal-vue@2.1.7
├─ qs@6.9.4
├─ query-string@6.13.1
│  ├─ decode-uri-component@^0.2.0
│  ├─ split-on-first@^1.0.0
│  └─ strict-uri-encode@^2.0.0
├─ queue-microtask@1.1.4
├─ readable-stream@3.6.0
│  ├─ inherits@^2.0.3
│  ├─ string_decoder@^1.1.1
│  └─ util-deprecate@^1.0.1
├─ safe-buffer@5.2.1
├─ semver@7.3.2
├─ split-on-first@1.1.0
├─ store@2.0.12
├─ strict-uri-encode@2.0.0
├─ string_decoder@1.3.0
│  └─ safe-buffer@~5.2.0
├─ string-hash@1.1.3
├─ superagent@5.3.1
│  ├─ component-emitter@^1.3.0
│  ├─ cookiejar@^2.1.2
│  ├─ debug@^4.1.1
│  ├─ fast-safe-stringify@^2.0.7
│  ├─ form-data@^3.0.0
│  ├─ formidable@^1.2.2
│  ├─ methods@^1.1.2
│  ├─ mime@^2.4.6
│  ├─ qs@^6.9.4
│  ├─ readable-stream@^3.6.0
│  └─ semver@^7.3.2
├─ supports-color@7.1.0
│  └─ has-flag@^4.0.0
├─ util-deprecate@1.0.2
├─ vue-functional-data-merge@3.1.0
├─ vue-router@3.5.1
├─ vue-upload-component@2.8.22
├─ vue@2.6.12
└─ vuex@3.6.2
```
</details>

The licenses generated from `yarn licenses list --prod`:

<details>
<summary>`yarn licenses list --prod`</summary>

```bash
yarn licenses v1.22.5
├─ (CC-BY-4.0 AND MIT)
│  ├─ @fortawesome/free-solid-svg-icons@5.14.0
│  │  ├─ URL: https://github.com/FortAwesome/Font-Awesome
│  │  ├─ VendorName: Dave Gandy
│  │  └─ VendorUrl: https://fontawesome.com/
│  └─ @fortawesome/free-solid-svg-icons@5.15.3
│     ├─ URL: https://github.com/FortAwesome/Font-Awesome
│     ├─ VendorName: Dave Gandy
│     └─ VendorUrl: https://fontawesome.com/
├─ Apache-2.0
│  └─ vue-upload-component@2.8.22
│     ├─ URL: git+https://github.com/lian-yue/vue-upload-component.git
│     ├─ VendorName: LianYue
│     └─ VendorUrl: https://github.com/lian-yue/vue-upload-component#readme
├─ BSD-3-Clause
│  ├─ column-resizer@1.3.5
│  │  ├─ URL: https://github.com/MonsantoCo/column-resizer.git
│  │  └─ VendorName: John Glynn
│  └─ qs@6.9.4
│     ├─ URL: https://github.com/ljharb/qs.git
│     └─ VendorUrl: https://github.com/ljharb/qs
├─ CC0-1.0
│  └─ string-hash@1.1.3
│     ├─ URL: git://github.com/darkskyapp/string-hash.git
│     └─ VendorName: The Dark Sky Company
├─ ISC
│  ├─ inherits@2.0.4
│  │  └─ URL: git://github.com/isaacs/inherits
│  └─ semver@7.3.2
│     └─ URL: https://github.com/npm/node-semver
├─ MIT
│  ├─ @fortawesome/fontawesome-common-types@0.2.30
│  │  ├─ URL: https://github.com/FortAwesome/Font-Awesome
│  │  ├─ VendorName: Dave Gandy
│  │  └─ VendorUrl: https://fontawesome.com/
│  ├─ @fortawesome/fontawesome-common-types@0.2.35
│  │  ├─ URL: https://github.com/FortAwesome/Font-Awesome
│  │  ├─ VendorName: Dave Gandy
│  │  └─ VendorUrl: https://fontawesome.com/
│  ├─ @fortawesome/fontawesome-svg-core@1.2.35
│  │  ├─ URL: https://github.com/FortAwesome/Font-Awesome
│  │  ├─ VendorName: Dave Gandy
│  │  └─ VendorUrl: https://fontawesome.com/
│  ├─ @fortawesome/vue-fontawesome@2.0.2
│  │  ├─ URL: https://github.com/FortAwesome/vue-fontawesome.git
│  │  └─ VendorUrl: https://github.com/FortAwesome/vue-fontawesome
│  ├─ @nuxt/opencollective@0.3.2
│  │  └─ URL: https://github.com/nuxt-contrib/opencollective.git
│  ├─ @triply/yasgui-utils@4.2.4
│  │  ├─ URL: https://github.com/TriplyDB/yasgui.git
│  │  ├─ VendorName: Triply
│  │  └─ VendorUrl: https://github.com/TriplyDB/yasgui
│  ├─ @triply/yasqe@4.2.7
│  │  ├─ URL: https://github.com/TriplyDB/yasgui.git
│  │  ├─ VendorName: Triply
│  │  └─ VendorUrl: https://github.com/TriplyDB/yasgui
│  ├─ @triply/yasr@4.2.8
│  │  ├─ URL: https://github.com/TriplyDB/yasgui.git
│  │  ├─ VendorName: Triply
│  │  └─ VendorUrl: https://github.com/TriplyDB/yasgui
│  ├─ @types/color-name@1.1.1
│  │  └─ URL: https://github.com/DefinitelyTyped/DefinitelyTyped.git
│  ├─ @types/lodash-es@4.17.3
│  │  └─ URL: https://github.com/DefinitelyTyped/DefinitelyTyped.git
│  ├─ @types/lodash@4.14.159
│  │  └─ URL: https://github.com/DefinitelyTyped/DefinitelyTyped.git
│  ├─ @types/node@14.0.23
│  │  └─ URL: https://github.com/DefinitelyTyped/DefinitelyTyped.git
│  ├─ ansi-styles@4.2.1
│  │  ├─ URL: https://github.com/chalk/ansi-styles.git
│  │  ├─ VendorName: Sindre Sorhus
│  │  └─ VendorUrl: sindresorhus.com
│  ├─ asynckit@0.4.0
│  │  ├─ URL: git+https://github.com/alexindigo/asynckit.git
│  │  ├─ VendorName: Alex Indigo
│  │  └─ VendorUrl: https://github.com/alexindigo/asynckit#readme
│  ├─ axios@0.21.1
│  │  ├─ URL: https://github.com/axios/axios.git
│  │  ├─ VendorName: Matt Zabriskie
│  │  └─ VendorUrl: https://github.com/axios/axios
│  ├─ bootstrap-vue@2.21.2
│  │  ├─ URL: https://github.com/bootstrap-vue/bootstrap-vue.git
│  │  └─ VendorUrl: https://bootstrap-vue.org/
│  ├─ bootstrap@4.6.0
│  │  ├─ URL: git+https://github.com/twbs/bootstrap.git
│  │  ├─ VendorName: The Bootstrap Authors
│  │  └─ VendorUrl: https://getbootstrap.com/
│  ├─ chalk@4.1.0
│  │  └─ URL: https://github.com/chalk/chalk.git
│  ├─ codemirror@5.56.0
│  │  ├─ URL: https://github.com/codemirror/CodeMirror.git
│  │  ├─ VendorName: Marijn Haverbeke
│  │  └─ VendorUrl: https://codemirror.net/
│  ├─ color-convert@2.0.1
│  │  ├─ URL: https://github.com/Qix-/color-convert.git
│  │  └─ VendorName: Heather Arthur
│  ├─ color-name@1.1.4
│  │  ├─ URL: git@github.com:colorjs/color-name.git
│  │  ├─ VendorName: DY
│  │  └─ VendorUrl: https://github.com/colorjs/color-name
│  ├─ colors@1.4.0
│  │  ├─ URL: http://github.com/Marak/colors.js.git
│  │  ├─ VendorName: Marak Squires
│  │  └─ VendorUrl: https://github.com/Marak/colors.js
│  ├─ combined-stream@1.0.8
│  │  ├─ URL: git://github.com/felixge/node-combined-stream.git
│  │  ├─ VendorName: Felix Geisendörfer
│  │  └─ VendorUrl: https://github.com/felixge/node-combined-stream
│  ├─ commander@5.1.0
│  │  ├─ URL: https://github.com/tj/commander.js.git
│  │  └─ VendorName: TJ Holowaychuk
│  ├─ component-emitter@1.3.0
│  │  └─ URL: https://github.com/component/emitter.git
│  ├─ consola@2.15.3
│  │  └─ URL: https://github.com/nuxt/consola.git
│  ├─ cookiejar@2.1.2
│  │  ├─ URL: https://github.com/bmeck/node-cookiejar.git
│  │  └─ VendorName: bradleymeck
│  ├─ core-js@3.9.1
│  │  └─ URL: https://github.com/zloirock/core-js.git
│  ├─ datatables.net-dt@1.10.21
│  │  ├─ URL: https://github.com/DataTables/Dist-DataTables-DataTables.git
│  │  ├─ VendorName: SpryMedia Ltd
│  │  └─ VendorUrl: https://datatables.net/
│  ├─ datatables.net@1.10.21
│  │  ├─ URL: https://github.com/DataTables/Dist-DataTables.git
│  │  ├─ VendorName: SpryMedia Ltd
│  │  └─ VendorUrl: https://datatables.net/
│  ├─ debug@4.1.1
│  │  ├─ URL: git://github.com/visionmedia/debug.git
│  │  └─ VendorName: TJ Holowaychuk
│  ├─ decode-uri-component@0.2.0
│  │  ├─ URL: https://github.com/SamVerschueren/decode-uri-component.git
│  │  ├─ VendorName: Sam Verschueren
│  │  └─ VendorUrl: github.com/SamVerschueren
│  ├─ delayed-stream@1.0.0
│  │  ├─ URL: git://github.com/felixge/node-delayed-stream.git
│  │  ├─ VendorName: Felix Geisendörfer
│  │  └─ VendorUrl: https://github.com/felixge/node-delayed-stream
│  ├─ fast-safe-stringify@2.0.7
│  │  ├─ URL: git+https://github.com/davidmarkclements/fast-safe-stringify.git
│  │  ├─ VendorName: David Mark Clements
│  │  └─ VendorUrl: https://github.com/davidmarkclements/fast-safe-stringify#readme
│  ├─ follow-redirects@1.13.0
│  │  ├─ URL: git@github.com:follow-redirects/follow-redirects.git
│  │  ├─ VendorName: Ruben Verborgh
│  │  └─ VendorUrl: https://github.com/follow-redirects/follow-redirects
│  ├─ form-data@3.0.0
│  │  ├─ URL: git://github.com/form-data/form-data.git
│  │  ├─ VendorName: Felix Geisendörfer
│  │  └─ VendorUrl: http://debuggable.com/
│  ├─ formidable@1.2.2
│  │  ├─ URL: https://github.com/node-formidable/formidable.git
│  │  └─ VendorUrl: https://github.com/node-formidable/formidable
│  ├─ has-flag@4.0.0
│  │  ├─ URL: https://github.com/sindresorhus/has-flag.git
│  │  ├─ VendorName: Sindre Sorhus
│  │  └─ VendorUrl: sindresorhus.com
│  ├─ jquery@3.5.1
│  │  ├─ URL: https://github.com/jquery/jquery.git
│  │  ├─ VendorName: JS Foundation and other contributors
│  │  └─ VendorUrl: https://jquery.com/
│  ├─ json2csv@5.0.1
│  │  ├─ URL: https://github.com/zemirco/json2csv
│  │  ├─ VendorName: Mirco Zeiss
│  │  └─ VendorUrl: http://zemirco.github.io/json2csv
│  ├─ jsonparse@1.3.1
│  │  ├─ URL: http://github.com/creationix/jsonparse.git
│  │  └─ VendorName: Tim Caswell
│  ├─ lodash-es@4.17.15
│  │  ├─ URL: https://github.com/lodash/lodash.git
│  │  ├─ VendorName: John-David Dalton
│  │  └─ VendorUrl: https://lodash.com/custom-builds
│  ├─ lodash.get@4.4.2
│  │  ├─ URL: https://github.com/lodash/lodash.git
│  │  ├─ VendorName: John-David Dalton
│  │  └─ VendorUrl: https://lodash.com/
│  ├─ methods@1.1.2
│  │  └─ URL: https://github.com/jshttp/methods.git
│  ├─ mime-db@1.44.0
│  │  └─ URL: https://github.com/jshttp/mime-db.git
│  ├─ mime-types@2.1.27
│  │  └─ URL: https://github.com/jshttp/mime-types.git
│  ├─ mime@2.4.6
│  │  ├─ URL: https://github.com/broofa/mime
│  │  ├─ VendorName: Robert Kieffer
│  │  └─ VendorUrl: http://github.com/broofa
│  ├─ ms@2.1.2
│  │  └─ URL: https://github.com/zeit/ms.git
│  ├─ n3@1.6.2
│  │  ├─ URL: https://github.com/rdfjs/N3.js.git
│  │  └─ VendorName: Ruben Verborgh
│  ├─ node-fetch@2.6.1
│  │  ├─ URL: https://github.com/bitinn/node-fetch.git
│  │  ├─ VendorName: David Frank
│  │  └─ VendorUrl: https://github.com/bitinn/node-fetch
│  ├─ popper.js@1.16.1
│  │  ├─ URL: git+https://github.com/FezVrasta/popper.js.git
│  │  ├─ VendorName: Federico Zivolo
│  │  └─ VendorUrl: https://popper.js.org/
│  ├─ portal-vue@2.1.7
│  │  ├─ URL: https://github.com/LinusBorg/portal-vue
│  │  ├─ VendorName: Thorsten Lünborg
│  │  └─ VendorUrl: htps://github.com/LinusBorg/
│  ├─ query-string@6.13.1
│  │  ├─ URL: https://github.com/sindresorhus/query-string.git
│  │  ├─ VendorName: Sindre Sorhus
│  │  └─ VendorUrl: https://sindresorhus.com
│  ├─ queue-microtask@1.1.4
│  │  ├─ URL: git://github.com/feross/queue-microtask.git
│  │  ├─ VendorName: Feross Aboukhadijeh
│  │  └─ VendorUrl: https://github.com/feross/queue-microtask
│  ├─ readable-stream@3.6.0
│  │  └─ URL: git://github.com/nodejs/readable-stream
│  ├─ safe-buffer@5.2.1
│  │  ├─ URL: git://github.com/feross/safe-buffer.git
│  │  ├─ VendorName: Feross Aboukhadijeh
│  │  └─ VendorUrl: https://github.com/feross/safe-buffer
│  ├─ split-on-first@1.1.0
│  │  ├─ URL: https://github.com/sindresorhus/split-on-first.git
│  │  ├─ VendorName: Sindre Sorhus
│  │  └─ VendorUrl: sindresorhus.com
│  ├─ store@2.0.12
│  │  ├─ URL: git://github.com/marcuswestin/store.js.git
│  │  ├─ VendorName: Marcus Westin
│  │  └─ VendorUrl: https://github.com/marcuswestin/store.js#readme
│  ├─ strict-uri-encode@2.0.0
│  │  ├─ URL: https://github.com/kevva/strict-uri-encode.git
│  │  ├─ VendorName: Kevin Mårtensson
│  │  └─ VendorUrl: github.com/kevva
│  ├─ string_decoder@1.3.0
│  │  ├─ URL: git://github.com/nodejs/string_decoder.git
│  │  └─ VendorUrl: https://github.com/nodejs/string_decoder
│  ├─ superagent@5.3.1
│  │  ├─ URL: git://github.com/visionmedia/superagent.git
│  │  ├─ VendorName: TJ Holowaychuk
│  │  └─ VendorUrl: https://github.com/visionmedia/superagent
│  ├─ supports-color@7.1.0
│  │  ├─ URL: https://github.com/chalk/supports-color.git
│  │  ├─ VendorName: Sindre Sorhus
│  │  └─ VendorUrl: sindresorhus.com
│  ├─ util-deprecate@1.0.2
│  │  ├─ URL: git://github.com/TooTallNate/util-deprecate.git
│  │  ├─ VendorName: Nathan Rajlich
│  │  └─ VendorUrl: https://github.com/TooTallNate/util-deprecate
│  ├─ vue-functional-data-merge@3.1.0
│  │  ├─ URL: git+https://github.com/alexsasharegan/vue-functional-data-merge.git
│  │  ├─ VendorName: Alex Regan
│  │  └─ VendorUrl: https://github.com/alexsasharegan/vue-functional-data-merge#readme
│  ├─ vue-router@3.5.1
│  │  ├─ URL: https://github.com/vuejs/vue-router.git
│  │  ├─ VendorName: Evan You
│  │  └─ VendorUrl: https://github.com/vuejs/vue-router#readme
│  ├─ vue@2.6.12
│  │  ├─ URL: git+https://github.com/vuejs/vue.git
│  │  ├─ VendorName: Evan You
│  │  └─ VendorUrl: https://github.com/vuejs/vue#readme
│  └─ vuex@3.6.2
│     ├─ URL: git+https://github.com/vuejs/vuex.git
│     ├─ VendorName: Evan You
│     └─ VendorUrl: https://github.com/vuejs/vuex#readme
└─ MPL-2.0 OR Apache-2.0
   └─ dompurify@2.0.12
      ├─ URL: git://github.com/cure53/DOMPurify.git
      ├─ VendorName: Mario Heiderich
      └─ VendorUrl: https://github.com/cure53/DOMPurify
```
</details>

I used `--prod` to show only what would be shipped with Fuseki assets (the dev is used only for building).

@afs do you think we should generate that output during build, save it somewhere, include in the final archive and amend our NOTICE saying something like "Fuseki UI is a Vue application that includes JavaScript libraries with licenses listed in the file $blabla" ?

(will reply your other notes in separate comments)

Thanks!
Bruno

kinow commented:
@afs 

>How easy is to to provide a version with no "manage" version?

Depends on how we want to do it.

1) We can simply comment the menu entry, leaving the rest of the code still accessible if the user knows the URL.

![image](https://user-images.githubusercontent.com/304786/143722072-b87d0f8a-5f2b-4c28-96f7-c46f4e7da756.png)

2) We could have a `prop` in the `Menu` component (a `prop` is like a parameter passed by the caller of the component) that enables/disables the "Manage" link.

```diff
diff --git a/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue b/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
index 316d913e63..d36876f133 100644
--- a/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
+++ b/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
@@ -26,7 +26,7 @@
         <b-nav-item to="/" exact-active-class="active">
           <FontAwesomeIcon icon="database" /> datasets
         </b-nav-item>
-        <b-nav-item to="/manage" active-class="active">
+        <b-nav-item v-if="isManageEnabled" to="/manage" active-class="active">
           <FontAwesomeIcon icon="cogs" /> manage
         </b-nav-item>
         <b-nav-item to="/documentation" active-class="active">
@@ -51,6 +51,13 @@ library.add(faDatabase, faCogs, faInfoCircle)
 export default {
   name: 'Menu',
 
+  props: {
+    isManageEnabled: {
+      type: Boolean,
+      default: false
+    }
+  },
+
   components: {
     FontAwesomeIcon,
     ServerStatus
```

3) There could also be a toggle somewhere (set either during build or runtime by the user, or even loaded from SPARQL query, etc) to control the manage. We would just have to control when routes are enabled or not (these VueRouter routes are liked Servlets mappings).

e.g.

```diff
diff --git a/jena-fuseki2/jena-fuseki-ui/src/router/index.js b/jena-fuseki2/jena-fuseki-ui/src/router/index.js
index f4b3e2ff66..8805f17207 100644
--- a/jena-fuseki2/jena-fuseki-ui/src/router/index.js
+++ b/jena-fuseki2/jena-fuseki-ui/src/router/index.js
@@ -21,6 +21,14 @@ import Home from '../views/Home'
 
 Vue.use(VueRouter)
 
+function checkManageEnabled (to, from, next) {
+  // FIXME: add a toggle somewhere to enable/disable manage feature
+  // eslint-disable-next-line no-constant-condition
+  if (false) {
+    next()
+  }
+}
+
 const routes = [
   {
     path: '/',
@@ -54,17 +62,20 @@ const routes = [
   {
     path: '/manage',
     name: 'ManageDatasets',
-    component: () => import(/* webpackChunkName: "manageDatasets" */ '../views/manage/ExistingDatasets')
+    component: () => import(/* webpackChunkName: "manageDatasets" */ '../views/manage/ExistingDatasets'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/manage/new',
     name: 'NewDataset',
-    component: () => import(/* webpackChunkName: "newDataset" */ '../views/manage/NewDataset')
+    component: () => import(/* webpackChunkName: "newDataset" */ '../views/manage/NewDataset'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/manage/tasks',
     name: 'Tasks',
-    component: () => import(/* webpackChunkName: "tasks" */ '../views/manage/Tasks')
+    component: () => import(/* webpackChunkName: "tasks" */ '../views/manage/Tasks'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/documentation',
```

4) Finally, I think we could also have a separate application, and break the UI into modules. One application fuseki-ui-core & fuseki-ui-manage, etc; another application loading just `fuseki-ui-core` views&routes, etc. A bit more complicated to make this one, but in the long run could give more flexibility on what can be done.

WDYT @afs? Happy to go with the simplest approach now, improve later, or in any other way (it'll be fun doing that anyway :) )

kinow commented:
@afs 

>How easy is to to provide a version with no "manage" version?

Depends on how we want to do it.

1) We can simply comment the menu entry, leaving the rest of the code still accessible if the user knows the URL.

![image](https://user-images.githubusercontent.com/304786/143722072-b87d0f8a-5f2b-4c28-96f7-c46f4e7da756.png)

2) We could have a `prop` in the `Menu` component (a `prop` is like a parameter passed by the caller of the component) that enables/disables the "Manage" link.

```diff
diff --git a/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue b/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
index 316d913e63..d36876f133 100644
--- a/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
+++ b/jena-fuseki2/jena-fuseki-ui/src/components/Menu.vue
@@ -26,7 +26,7 @@
         <b-nav-item to="/" exact-active-class="active">
           <FontAwesomeIcon icon="database" /> datasets
         </b-nav-item>
-        <b-nav-item to="/manage" active-class="active">
+        <b-nav-item v-if="isManageEnabled" to="/manage" active-class="active">
           <FontAwesomeIcon icon="cogs" /> manage
         </b-nav-item>
         <b-nav-item to="/documentation" active-class="active">
@@ -51,6 +51,13 @@ library.add(faDatabase, faCogs, faInfoCircle)
 export default {
   name: 'Menu',
 
+  props: {
+    isManageEnabled: {
+      type: Boolean,
+      default: false
+    }
+  },
+
   components: {
     FontAwesomeIcon,
     ServerStatus
```

3) There could also be a toggle somewhere (set either during build or runtime by the user, or even loaded from SPARQL query, etc) to control the manage. We would just have to control when routes are enabled or not (these VueRouter routes are liked Servlets mappings).

e.g.

```diff
diff --git a/jena-fuseki2/jena-fuseki-ui/src/router/index.js b/jena-fuseki2/jena-fuseki-ui/src/router/index.js
index f4b3e2ff66..8805f17207 100644
--- a/jena-fuseki2/jena-fuseki-ui/src/router/index.js
+++ b/jena-fuseki2/jena-fuseki-ui/src/router/index.js
@@ -21,6 +21,14 @@ import Home from '../views/Home'
 
 Vue.use(VueRouter)
 
+function checkManageEnabled (to, from, next) {
+  // FIXME: add a toggle somewhere to enable/disable manage feature
+  // eslint-disable-next-line no-constant-condition
+  if (false) {
+    next()
+  }
+}
+
 const routes = [
   {
     path: '/',
@@ -54,17 +62,20 @@ const routes = [
   {
     path: '/manage',
     name: 'ManageDatasets',
-    component: () => import(/* webpackChunkName: "manageDatasets" */ '../views/manage/ExistingDatasets')
+    component: () => import(/* webpackChunkName: "manageDatasets" */ '../views/manage/ExistingDatasets'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/manage/new',
     name: 'NewDataset',
-    component: () => import(/* webpackChunkName: "newDataset" */ '../views/manage/NewDataset')
+    component: () => import(/* webpackChunkName: "newDataset" */ '../views/manage/NewDataset'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/manage/tasks',
     name: 'Tasks',
-    component: () => import(/* webpackChunkName: "tasks" */ '../views/manage/Tasks')
+    component: () => import(/* webpackChunkName: "tasks" */ '../views/manage/Tasks'),
+    beforeEnter: checkManageEnabled
   },
   {
     path: '/documentation',
```

4) Finally, I think we could also have a separate application, and break the UI into modules. One application fuseki-ui-core & fuseki-ui-manage, etc; another application loading just `fuseki-ui-core` views&routes, etc. A bit more complicated to make this one, but in the long run could give more flexibility on what can be done.

WDYT @afs? Happy to go with the simplest approach now, improve later, or work on a more complex solution if that'd be better too (it'll be fun anyway :) )

kinow commented:
>As it depends on no Jena artifacts, can we move the build to just after jena-fuseki-access? This makes it "one of the modules" and can then go into Fuseki Mian as easily as Fuseki webapp.

Hmmm, I think I will need to go back to the whiteboard to see how the modules work in Jena. But it should be doable I guess. The new `jena-fuseki-ui` uses Maven to call `yarn build`. The generated resources (HTML, JS, CSS, images) can be used as-is in a separate HTTP server if we wanted. By default they point to URL's like `/$/ping` I think (so if behind a reverse proxy they should work as long as the URL is correct for Fuseki servlets).

So as long as the final result includes these generated resources somewhere where Fuseki server can load when expanding the WAR or loading the Jetty application, there's no constraint as how the Maven project is organized.

i.e. I just need to learn how to achieve what you explained, but should work just fine :crossed_fingers:

kinow commented:
>Every time I going to the query tab it seems quite slow (but my machine may not be setup correctly having flipped UI implementations and general have a polluted browser).

Hmmm, just noticed that. I enabled throttling in the browser development tools. I must be loading something unnecessarily, or blocking the browser's main thread and preventing the UI from rendering. Will investigate :+1:

kinow commented:
>Formatting of results - can have have "" on strings and _: for bnode labels (I realise this is about yasgui/yasr)

Maybe a follow-up issue. I think the development in yasgui/yasr is slow/stopped. But being JS, we can try to tap into the yasgui/yasr objects when loading the JS app, and change a little bit how they behave.

So in other words, we can have fixes for bugs in yasgui/yasr in Fuseki UI. Then we submit pull requests there and wait until merged, and remove our patched code once it's done.

kinow commented:
BTW, if anyone would like to try changing the Fuseki UI code, just start `FusekiServer` either from jar or Eclipse. The important part is to have `localhost:3030` serving Fuseki.

Then start the UI in development mode. By going to `fuseki2/jena-fuseki-ui` directory, installing dependencies with `yarn install`, and start the dev server with `yarn run serve`.

Now change the code as much as you like. The `run serve` command starts a dev server that monitors for changes in the source code (styles, JS, HTML/templates, etc), and re-starts the dev server in ms/seconds.

The dev server will be available in `http://localhost:8080` (important, not 3030, but 8080). It has a `devServer` proxy (see the `fuseki-ui` configuration files in the root) which proxies requests to `:8080/$/ping` to `:3030/$/ping` (hence the requirement to have Fuseki Server listening 3030 in development).

The port number shouldn't matter in production or when the UI is packaged.

kinow commented:
On the performance, the framerate is about 60 fps, then if I remove the code that loads the query editors (yasqe) then it keeps the framerate about the same.

![Screenshot from 2021-11-28 22-33-53](https://user-images.githubusercontent.com/304786/143789024-e93de3c7-5da0-4467-9866-76a93bced775.png)

But leaving just this:

```js
      vm.yasqe = new Yasqe(
        document.getElementById('yasqe'),
        {
          showQueryButton: true,
          resizeable: false,
          requestConfig: {
            endpoint: `/${vm.datasetName}/sparql`
          }
        }
      )
```

It drops from 60 to under 10, since the main thread is busy creating that object, and the impression that the UI is really slow when opening the query editor.

To improve user experience a little, I have deferred the YASR/YASQE DOM operations, and added a spinner/loading indication in the UI. Also fixed a bug where I used the wrong prop in the `TableListing` component (re-used whenever we render a table, like listing datasets), so now the `TableListing` components should display a spinner/loading symbol too.

The fps is still dropping, but not much I can do I think, except for trying to understand what can be improved in YASR/YASQE. These libraries manipulate the DOM, creating components, attaching to the DOM, resizing it, adding listeners, etc.

![Screenshot from 2021-11-29 11-26-28](https://user-images.githubusercontent.com/304786/143789111-4b4a7237-a7bc-4320-85e3-a20529c731a9.png)

(In the image above the fps drop is still present, but shouldn't bother users as much as before).

@afs one item fixed I think :+1:

rvesse commented:
If you've pushed the YASQE loading onto a background thread and it still tanks performance a bit there's not a lot more you can do about it other than talk to the YASQE people and see if there's performance improvements to be made

rvesse commented:
The `props` style approach can work quite well for these kinds of UIs.  Haven't worked with Vue personally but we have some React based UIs in $dayjob. I know one limitation of this approach is that the `props` (or `configuration` in React terminology) is typically baked in at `yarn build` time so potentially tricky to toggle on/off at runtime.

But perhaps Vue is more flexible in this regard?

Definitely something to consider in choosing what development direction to pursue

kinow commented:
@rvesse 

>If you've pushed the YASQE loading onto a background thread and it still tanks performance a bit there's not a lot more you can do about it other than talk to the YASQE people and see if there's performance improvements to be made

+1 I had a look at the code, found a few settings and tried toggling the values to see if it would make any difference (maybe disabling something we don't need in Fuseki UI), but it didn't help much. For a follow-up I would definitely investigate a few performance improvements to recommend to YASQE (e.g. simplifying the structure in the DOM, or pre-creating elements, etc).

>I know one limitation of this approach is that the props (or configuration in React terminology) is typically baked in at yarn build time so potentially tricky to toggle on/off at runtime.
>But perhaps Vue is more flexible in this regard?

I think Vue and React are pretty similar in this aspect. But we can still:

- load a value from the user browser (cookie, local storage, etc)
- load the value from a configuration setting passed to fuseki during start-up, and passed down to the UI
- load the value dynamically from something like a Servlet in Fuseki that reads from disk, from a dataset, or from some other service, etc

And then just pass that value down via a prop to the component.

afs commented:
I'm tracking down changes in the apache-jena-fuseki binary. It has grown from 62M to 70M (zip) and 70M to 83M unpacked. Maybe it is just what it is. We can separately discuss whether shipping a war file lik this at all is necessary nowadays and whether we could just put a link in the download page (and the README for apache-jena-fuseki) to get from maven central.

The size growth is probably my mistake somewhere. 

jena-fuseki-ui/node_modules/ is 580M and the list from "yarn list --prod" is 150 items. Could that be right or have I contaminated my setup?

Current:
```
5.0M    apache-jena-fuseki-old/webapp
30M	apache-jena-fuseki-old/fuseki.war
35M	apache-jena-fuseki-old/fuseki-server.jar
```
New UI:
```
16M	apache-jena-fuseki-ui/webapp
33M	apache-jena-fuseki-ui/fuseki.war
35M	apache-jena-fuseki-ui/fuseki-server.jar
```
One js.map files are over 4M and one is 2M. They are in the system more than once -- that's not wrong, it's a consequence war file and fuseki-server.jar+webapp.

afs commented:
vue / react: I'm not qualified to comment. In large enterprise settings, I almost always see react. In less traditional settings, vue comes up.

And we have a vue implementation here already :clap:.

Is this for 4.3.0 for Fuseki/webapp?
Or put it in just after 4.3.0 and ask people to try it out?
(making a workbench only, query+upload data, is a separate matter for now.)

Critical path tasks:

- [ ] Confirm the size and build is right.
- [ ] LICENCE and NOTICE for code
- [ ] LICENCE and NOTICE for war binary and zip file
- [ ] ???

kinow commented:
@afs 

>Confirm the size and build is right.

It looks correct to me. The fontawesome dependencies are bringing a few more transitive dependencies. We could possibly limit the icons used, and import them directly as SVG's, which would possibly reduce the package size. But YASQE/YASGUI/YASR were updated and now require a lot more dependencies than the version used by Fuseki in the past. Finally, Bootstrap Vue also includes a few more dependencies. We could see about depending directly on Bootstrap, instead of Bootstrap Vue, but before that we would have to write some components ourselves.

So there's room for improvement regarding the artefact size, if that becomes a problem. But I don't think you have any issues in your environment, I think your analysis is correct :+1: 

>vue / react: I'm not qualified to comment. In large enterprise settings, I almost always see react. In less traditional settings, vue comes up.
>
>And we have a vue implementation here already

Yeah, React is the most famous JS reactive framework, mostly because it was created first. It has more libraries and documentation, but as I was working with Vue, and since Fuseki is a very simple application with few dependencies, I wrote it in Vue to save time.

After we have it merged, my plans are:

- [ ] write unit and end-to-end tests to cover most basic functionality (formatting a query, submitting a query, clicking on the menu to navigate the site, etc)
- [ ] move the common code (i.e. not specific to vue) to JS or TypeScript code. So that we can always use this code, no matter the framework (I migrated a lot of old JS code that was in the middle of Backbone.js code; now this code is in the middle of some Vue code/components; IMHO moving it into a separate place will make maintenance and future upgrades like this a lot smoother)
- [ ] see which direction we will take, adding more features? breaking it into other webapps like manage/sparqler/etc? improve performance? etc, and then create new JIRA issues for that.

>Is this for 4.3.0 for Fuseki/webapp?
>Or put it in just after 4.3.0 and ask people to try it out?
>(making a workbench only, query+upload data, is a separate matter for now.)

Up to you @afs. I think the sooner we give it to users to try it, the better. But I don't fully understand the build and deployment of Fuseki, so cannot say I'm 100% confident I have done a good job packaging it.

> LICENCE and NOTICE for code

Uh! I forgot about the `LICENSE` file for `jena-fuseki-ui`! Just included one with the Apache License at the top, followed by a separator `- - - - - - - - - - - - - - - - - - - - - - -`, and the output of `yarn licenses list --prod`. And added a `NOTICE` file with the output of `yarn licenses generate-disclaimer --prod`. I included in comments the commands used to generate the files so we remember to update them when changing the dependencies of the projects.

> LICENCE and NOTICE for war binary and zip file

I just had a look at `apache-jena-fuseki` module's assembly configuration. It already includes a `LICENSE` file and a `NOTICE` file. The `LICENSE` file doesn't mention "bootstrap" (a dependency we used in the existing Fuseki code), so I did not touch that file. The `NOTICE` file of `apache-jena-fuseki`, though, mentions bootstrap.

I replaced the section with dependencies that appear to refer to the JS/CSS libraries in the existing Fuseki UI, replacing the the licenses from the `jena-fuseki-ui` module.

>???

Only thing I can remember now that I wanted to mention before, is that this pull request does not delete the old Fuseki files. They are included in the source and distributed code. However, the new Fuseki UI files are copied to the distributed ZIP, and the `index.html` is overwritten, using the new JS/CSS files.

Thanks @afs !

afs commented:
If that's the size it is, and my build isn't messed up, that's OK. Getting "smart" costs the project in maintaining the smarts. There are bigger download savings to made like distributing the WAR file by link from the download page, not in the zip file.

4.3.0 is a lot of internal churn for networking changes and reworking things to make local dataset and remote dataset have the same API + using java.net.http, various fixes, retires OSGi formally, not much external features other than xloader. For Fuseki users, not much. But the misc internal improvements may not be perfect and then noise around that distracts from the great new UI. It is inevitable the usual suspects will comment.

From the comments above, we have some sorting out to do, not a huge amount but some. That's OK.

I'd prefer not do a release too close to Christmas because of any follow-up and users@ traffic.

Proposal - if this is OK, I'll write to dev@.

- 4.3.0 ASAP as-is. It's ready to go; I just ran all the builds.
- merge this PR a few days after - or maybe start a shared branch in our repo and work there?
- Aim for 4.4.0 in January. I can find time to do a release, and I assume Bruno will vote for it :smile: if it works so that's 2 possible votes. It would be, at a minimum, the work so far with the LICENSE and NOTICE sorted out.

There are various other Fuseki things I can think of but not that should be critical path. Something else for dev@ and it might be worth starting a JIRA epic.

kinow commented:
+1 @afs, thanks!

afs commented:
Minor news item: I've prototyped the admin functions (create, delete at `/$/datasets` and backups) as a Fuseki Module for Fuseki Main and secured them for localhost only.

Fuskei Main does not have Shiro (it can work with Shiro; for users/password it uses the Jetty one).

For securing admin generally, we know localhost only is inconvenient at times. I would still not like "open" as the default. I'm wondering about command line arg: `--admin=FILE` where FILE is JSON `{ user: "", password: ""}` (so the password isn't on the command line and viewable with process viewing tools) which would allow it from anywhere.  Maybe a special `--admin=*` for explicitly no user/password for quick-convenience. At least then it is an explicit step by the user.

afs commented:
Let's see if we can move this closer to being merged into main.

I'd like to leave main untouched for a few more days just in case there are more security issues but Jena as released is on log4j2 2.16.0 now. The latest issue, fixed in 2.17.0, does not affect Jena unless the user changes the logging or has a compromised machine. So good to upgrade Jena's dependency, but from the current information, it is not a reason to do a jena security release. That might be revised - we'll see.

With a goal of 4.4.0 with new UI for Fuseki/webapp (war file and fulljar),

* Rebase branch to current main - there have been POM changes but visually looking at this PR I don't see any conflicts.
* Remove unused. Delete the old UI and any other code. We can get it back from git in the unlikely case we need it.
* Squash the intermediate commits to a few key commits for the long term log record. One commit "New UI" would work for me.

The change to having no UI in jena-fuseki-webapp will change development a bit (for me - running in a debugger off the codebase directly) but jena-fuseki-webapp isn't changing rapidly and it can be sorted out later after integration into main.

@kinow Thank for your patience!

kinow commented:
- [x] Rebase branch to current main - there have been POM changes but visually looking at this PR I don't see any conflicts.
- [ ] Remove unused. Delete the old UI and any other code. We can get it back from git in the unlikely case we need it.
- [ ] Squash the intermediate commits to a few key commits for the long term log record. One commit "New UI" would work for me.

kinow commented:
@afs I updated your requirements to have a bullet item list on GH, and went through each. I think I got everything, but not sure if I deleted all the files for the old UI, or if there were more files to delete.

>@kinow Thank for your patience!

Thank **you** for your patience too, and for the latest releases :+1: :christmas_tree:

afs commented:
Great. I see lots of old files removed; I don't think there are others in unusual place.

kinow commented:
:tada: @afs!

>When this PR and the other code-changing PRs go to main

I don't remember the process now, should I merge after this has been approved, or should I wait for someone else to merge it?

afs commented:
Merge away!

I pulled your branch, built the codebase, ran fuseki-server, created a DB in UI, added data and queried it.
